File: Combination Sum.cpp
class Solution {
   public:
    void make_comb(vector<vector<int>>& ans, vector<int>& now, const vector<int>& nums, int index, int cur, int target) {
        if (cur == target) {
            ans.push_back(now);
            return;
        }
        if (cur > target) return;
        for (int i = index; i < nums.size(); ++i) {
            now.push_back(nums[i]);
            make_comb(ans, now, nums, i, cur + nums[i], target);
            now.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> now;
        make_comb(ans, now, candidates, 0, 0, target);
        return ans;
    }
};

File: Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def solve(ans_list, numbers, target, idx, val_list):
            if target < 0:
                return
            if target == 0:
                ans_list.append(copy.copy(val_list))
                return
            for i in range(idx, len(numbers)):
                num = numbers[i]
                val_list.append(num)
                solve(ans_list, numbers, target - num, i, val_list)
                val_list.pop()
        ans_list = []
        solve(ans_list, candidates, target, 0, [])
        return ans_list
